### PR TITLE
ENH: SPMD interface for IncrementalLinearRegression

### PR DESCRIPTION
**Pull request by olegkkruglov**
### Description

* Added SPMD interface for IncrementalLinearRegression
* Changed policy saving workflow, now queue is saved to attributes instead of policy. It is necessary because `finalize_fit` requires `spmd_policy`, but `partial_fit` requires `data_parallel_policy` on oneDAL side
* `finalize_fit` now uses provided queue for computations on onedal4py side. 
* Contains some content from https://github.com/intel/scikit-learn-intelex/pull/1777 for test implementation

---

Checklist to comply with before moving PR from draft:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.  
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.

### CODE REVIEW

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import numpy as np
+
+from daal4py.sklearn._utils import get_dtype
+
+from ..._device_offload import support_usm_ndarray
+from ...common.hyperparameters import get_hyperparameters
+from ...datatypes import _convert_to_supported, to_table
+from ...linear_model import (
+    IncrementalLinearRegression as IncrementalLinearRegression_nonSPMD,
```
Comment (ethanglaser) [inline onedal/spmd/linear_model/incremental_linear_model.py:25]:
```suggestion
    IncrementalLinearRegression as IncrementalLinearRegression_Batch,
```
Lets have uniformity across algos

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+)
+from ...utils import _check_X_y, _num_features
+from .._base import BaseEstimatorSPMD
+
+
+class IncrementalLinearRegression(BaseEstimatorSPMD, IncrementalLinearRegression_nonSPMD):
+    def _reset(self):
```
Comment (ethanglaser) [inline onedal/spmd/linear_model/incremental_linear_model.py:32]: Can you clarify what the point of _reset is? Seems like it just sets partial result?

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            self._params = self._get_onedal_params(self._dtype)
+
+        y = np.asarray(y).astype(dtype=self._dtype)
+        self._y_ndim_1 = y.ndim == 1
+
+        X, y = _check_X_y(X, y, dtype=[np.float64, np.float32], accept_2d_y=True)
+
+        self.n_features_in_ = _num_features(X, fallback_1d=True)
+        X_table, y_table = to_table(X, y)
+        hparams = get_hyperparameters("linear_regression", "train")
```
Comment (ethanglaser) [inline onedal/spmd/linear_model/incremental_linear_model.py:80]: Why is any of this needed? Shouldn't this be covered by batch estimator call? There are no data preprocessing function calls in other spmd functionality.

File: onedal/spmd/linear_model/incremental_linear_model.py (hunk 1 above)
Comment (olegkkruglov) [reply onedal/spmd/linear_model/incremental_linear_model.py:25]: as far as I understand, Incremental and Batch are contradicting properties, so the current naming is better for me. I'll change in other estimators.

File: onedal/spmd/linear_model/incremental_linear_model.py (hunk 2 above)
Comment (olegkkruglov) [reply onedal/spmd/linear_model/incremental_linear_model.py:32]: yes, it sets partial result to the clean one. it is called in sklearnex fit method.

File: onedal/spmd/linear_model/incremental_linear_model.py (hunk 3 above)
Comment (olegkkruglov) [reply onedal/spmd/linear_model/incremental_linear_model.py:80]: this method overrides the regular one, that's why data preprocessing is necessary here as well

File: onedal/spmd/linear_model/incremental_linear_model.py (hunk 1 above)
Comment (ethanglaser) [reply onedal/spmd/linear_model/incremental_linear_model.py:25]: Good point

**Comment by samir-nasibli** [comment]
@olegkkruglov please rebase your branch

File: onedal/linear_model/incremental_linear_model.py [hunk 4]
```diff
@@ -93,7 +94,7 @@ def partial_fit(self, X, y, queue=None):
         hparams = get_hyperparameters("linear_regression", "train")
         if hparams is not None and not hparams.is_default:
             self._partial_result = module.partial_train(
```
Comment (ethanglaser) [inline onedal/linear_model/incremental_linear_model.py:96]: is there a reason module is used here but deleted from other algos?

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        if hparams is not None and not hparams.is_default:
+            self._partial_result = module.partial_train(
```
Comment (ethanglaser) [inline onedal/spmd/linear_model/incremental_linear_model.py:82]: same question about module

File: onedal/linear_model/incremental_linear_model.py (hunk 4 above)
Comment (olegkkruglov) [reply onedal/linear_model/incremental_linear_model.py:96]: it is used here twice, that's why i kept it here for the code's brevity

File: onedal/spmd/linear_model/incremental_linear_model.py (hunk 5 above)
Comment (olegkkruglov) [reply onedal/spmd/linear_model/incremental_linear_model.py:82]: https://github.com/intel/scikit-learn-intelex/pull/1972#discussion_r1723074448

**Comment by olegkkruglov** [comment]
https://ecmd.jf.intel.com/commander/link/jobDetails/jobs/ef5f1fac-cd11-f152-9962-a4bf010d0e2e

**Comment by ethanglaser** [comment]
https://intel-ci.intel.com/ef5f1fac-cd11-f152-9962-a4bf010d0e2e

**Review by bdmoore1 — APPROVED** [review]

**Review by samir-nasibli — COMMENTED** [review]
@olegkkruglov more than one approval from the team is needed for all online algos merge.

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        policy = super(base_IncrementalLinearRegression, self)._get_policy(queue, X)
+
+        X, y = _convert_to_supported(policy, X, y)
+
+        if not hasattr(self, "_dtype"):
+            self._dtype = get_dtype(X)
+            self._params = self._get_onedal_params(self._dtype)
+
+        y = np.asarray(y).astype(dtype=self._dtype)
+        self._y_ndim_1 = y.ndim == 1
```
Comment (icfaust) [inline onedal/spmd/linear_model/incremental_linear_model.py:71]: https://github.com/intel/scikit-learn-intelex/pull/1957#discussion_r1708746789

There is a way to avoid using this _y_ndim_1 by using np.squeeze It was implemented in that PR and I would like for you to follow its implementation

**Review by icfaust — COMMENTED** [review]
Need to count finite checks in partial fit, there is a discussion of tolerances vs batch and a removal of an unnecessary variable.

Would it worth testing some of these versus the sklearnex standard linear regression (since its been battle-tested in use from almost the get-go of the repo)?

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+def test_incremental_linear_regression_fit_spmd_random(
+    dataframe, queue, fit_intercept, num_samples, num_features, macro_block, dtype
+):
+    # Import spmd and non-SPMD algo
+    from sklearnex.linear_model import IncrementalLinearRegression
+    from sklearnex.spmd.linear_model import (
+        IncrementalLinearRegression as IncrementalLinearRegression_SPMD,
+    )
+
+    tol = 1e-4 if dtype == np.float32 else 1e-7
```
[the remark this answers is not in the document]
Comment (olegkkruglov) [reply sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:210]: https://github.com/intel/scikit-learn-intelex/blob/e799ec43b876421c26a312264c242fe6fc27dbf7/sklearnex/linear_model/tests/test_incremental_linear.py#L187
yes, it is the same here

**Comment by ethanglaser** [comment]
Please resolve internal CI fails (both conformance and test threshold issues) before merging

**Comment by olegkkruglov** [comment]
https://intel-ci.intel.com/ef6abe1a-0992-f19f-bade-a4bf010d0e2e

File: onedal/utils/validation.py [hunk 8]
```diff
@@ -194,9 +194,13 @@ def _check_X_y(
...
     if not accept_2d_y:
         y = _column_or_1d(y, warn=True)
+    else:
+        y = np.ascontiguousarray(y)
+
     if y_numeric and y.dtype.kind == "O":
         y = y.astype(np.float64)
-    _assert_all_finite(y)
+    if force_all_finite:
+        _assert_all_finite(y)
```
Comment (ethanglaser) [inline onedal/utils/validation.py:204]: Rebase such that these changes do not show in this PR

File: onedal/spmd/linear_model/incremental_linear_model.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    API is the same as for `onedal.linear_model.IncrementalLinearRegression`.
+    """
+
+    def _reset(self):
+        self._partial_result = super(base_IncrementalLinearRegression, self)._get_backend(
+            "linear_model", "regression", "partial_train_result"
+        )
+
+    def partial_fit(self, X, y, queue=None):
```
Comment (ethanglaser) [inline onedal/spmd/linear_model/incremental_linear_model.py:42]: I am still a little confused why this is re-implemented and cannot take the base estimator's - can you clarify?

File: onedal/linear_model/incremental_linear_model.py [hunk 10]
```diff
@@ -113,36 +114,36 @@ def finalize_fit(self, queue=None):
...
-            )
+            result = module.finalize_train(policy, self._params, self._partial_result)
 
         self._onedal_model = result.model
 
         packed_coefficients = from_table(result.model.packed_coefficients)
         self.coef_, self.intercept_ = (
-            packed_coefficients[:, 1:],
-            packed_coefficients[:, 0],
+            packed_coefficients[:, 1:].squeeze(),
```
Comment (ethanglaser) [inline onedal/linear_model/incremental_linear_model.py:143]: Why would squeezing be needed here?

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+        hparams = inclin.get_hyperparameters("fit")
+        hparams.cpu_macro_block = macro_block
+        hparams.gpu_macro_block = macro_block
+
+        hparams_spmd = inclin_spmd.get_hyperparameters("fit")
+        hparams_spmd.cpu_macro_block = macro_block
+        hparams_spmd.gpu_macro_block = macro_block
+
+    inclin_spmd.fit(local_dpt_X, local_dpt_y)
+    inclin.fit(dpt_X, dpt_y)
```
Comment (ethanglaser) [inline sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:98]: Why is just fit() used here?

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+
+    assert_allclose(inclin.coef_, inclin_spmd.coef_, atol=tol)
+    if fit_intercept:
+        assert_allclose(inclin.intercept_, inclin_spmd.intercept_, atol=tol)
+
```
Comment (ethanglaser) [inline sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:242]: Does this test actually provide coverage for incremental algos or just base batch?

**Comment by ethanglaser** [comment]
CI looks good

File: onedal/linear_model/incremental_linear_model.py (hunk 10 above)
Comment (icfaust) [reply onedal/linear_model/incremental_linear_model.py:143]: Its a change related to this change, ideally IncrementalLinearRegression would be changed to follow changes that were requested in IncrementalRidge, but it looks like the tests were changed instead. https://github.com/intel/scikit-learn-intelex/pull/1972#discussion_r1741984101

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+@pytest.mark.skipif(
+    not _mpi_libs_and_gpu_available,
+    reason="GPU device and MPI libs required for test",
+)
+@pytest.mark.parametrize(
+    "dataframe,queue",
+    get_dataframes_and_queues(dataframe_filter_="dpnp,dpctl", device_filter_="gpu"),
+)
+@pytest.mark.parametrize("fit_intercept", [True, False])
+@pytest.mark.parametrize("num_blocks", [1, 2])
```
[the remark this answers is not in the document]
Comment (icfaust) [reply sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:114]: I would have liked to have seen a test along those lines for mathematical rigor, but it isn't vital for this PR.

**Review by icfaust — APPROVED** [review]
My approval is contingent on addressing @ethanglaser 's comments.

File: onedal/spmd/linear_model/incremental_linear_model.py (hunk 9 above)
Comment (olegkkruglov) [reply onedal/spmd/linear_model/incremental_linear_model.py:42]: we don't have spmd partial_fit on c++ side, that's why it is reimplemented here to take non-spmd backend

File: onedal/linear_model/incremental_linear_model.py (hunk 10 above)
Comment (icfaust) [reply onedal/linear_model/incremental_linear_model.py:143]: Thanks to @olegkkruglov my previous comment is very wrong. He has implemented the change that I mention in the step above.  As far as I understand, its because linear regression supports multi_output.  In the case that y is 1 dimensional, the onedal object may return a 2 dimensional object where the size in one of the dimensions is 1 and it must be squeezed out to match the original input y data.

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py (hunk 12 above)
Comment (olegkkruglov) [reply sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:242]: fit for incremental algos splits data into batches and then call partial_fit for each batch. it is done like in scikit incrementalPCA. this test provides coverage for spmd version of this method comparing it with regular IncrementalPCA.fit

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py (hunk 11 above)
Comment (olegkkruglov) [reply sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:98]: https://github.com/intel/scikit-learn-intelex/pull/1972#discussion_r1745114693

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py (hunk 12 above)
Comment (ethanglaser) [reply sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:242]: I figured, just couldn't find specifically where that logic was defined

File: onedal/linear_model/incremental_linear_model.py (hunk 10 above)
Comment (ethanglaser) [reply onedal/linear_model/incremental_linear_model.py:143]: But why would squeeze be needed in a test?

Comment (olegkkruglov) [reply onedal/linear_model/incremental_linear_model.py:143]: it was recognized after implementing this change that tests were not exactly correct, and after comparing with original sklearn, i've changed tests as well

File: sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+):
+    # Import spmd and non-SPMD algo
+    from sklearnex.linear_model import IncrementalLinearRegression
+    from sklearnex.spmd.linear_model import (
+        IncrementalLinearRegression as IncrementalLinearRegression_SPMD,
+    )
+
+    tol = 3e-4 if dtype == np.float32 else 1e-7
+
+    # Create gold data and process into dpt
```
Comment (ethanglaser) [inline sklearnex/spmd/linear_model/tests/test_incremental_linear_spmd.py:282]:
```suggestion
    # Generate random data and process into dpt
```

**Review by ethanglaser — APPROVED** [review]
